### PR TITLE
Handle non ASCII characters in decompiler widget.

### LINKED ISSUE #2358: Handle non-ASCII symbols in decompiler widget
**Is your feature request related to a problem? Please describe.**

Currently the effect of different  encodings on text offsets in the decompiler widget and RAnnotated code is completely ignored.
It is very likely that non ASCII symbols in the decompiler text will cause text offsets to shift.

**Describe the solution you'd like**

* Make sure that text encoding in decompiler plugins is what they claim it is and that offsets are correct
* Create the text position mapping between encoding used by `RAnnotatedCode` (supposedly UTF8) and whatever `QString` and `QTextCursor` uses (supposedly UTF16 based)

It will probably be easier to map all annotation positions after receiving data from the decompiler and be done with it instead of mapping back and forth for each interaction.

Carefully check the documentation and actual behavior of which text encoding exactly is used. Especially with UTF-16 there are plenty of libraries and programming languages using something very similar to UTF-16 but not exactly due to historical reasons.

When working with text offsets pay attention of what is being measured - byte position, character, code unit, graphemes or something else.

### PR DESCRIPTION
**Pull request by karliss**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**

Convert annotation offsets to positions Qt text edit. Wrong positions are obvious when using decompiler based highlighting but it also affects context menu interactions.

**Test plan (required)**

Before:
Screenshot from 2020-08-31 15-35-18
After:
Screenshot from 2020-08-31 14-36-50

**Closing issues**

Closes #2358

### CODE REVIEW

**Comment by XVilka** [comment]
Does it work with Arabic and Japanese?

**Comment by karliss** [comment]
> Does it work with Arabic and Japanese?

Feel free to try your self. I can't input or evaluated if they handled correctly. Arabic and other RTL scripts might have problems of their own.  Font issues is one more topic.

As can be seen in screenshot with emojis those are symbols outside Basic Multilingual Plane >0xffff. Which map to multiple QChars. The last symbol with familiy is made of multiple ones connected using zero width joiner. That covers both ends of Unicode  so there shouldn't be too much issues with anything in the middle, as good as Qt supports it. Wanted to try some of the CJK symbols > 0xffff but I didn't have appropriate fonts setup on my system.

Forgot to mention that when I tried with retdec I couldn't properly test it because json to RAnnotated code conversion isn't merged yet so highlighting wasn't visible. There were also some json decoding isues when using symbols > 0xffff but those were from earlier steps before decompiler widget.

**Comment by ITAYC0HEN** [comment]
Hebrew: first comment RTL, second is LTR
image

Indeed, I agree that this is another issue out of scope

**Review by ITAYC0HEN — APPROVED** [review]

File: src/widgets/DecompilerWidget.cpp [hunk 1]
```diff
@@ -565,14 +565,56 @@ bool DecompilerWidget::addressInRange(RVA addr)
...
+    QTextStream stream(bytes);
+    stream.setCodec("UTF-8");
+    std::vector<size_t> offsets;
+    offsets.reserve(bytes.size());
+    offsets.push_back(0);
+    QChar singleChar;
+    while (!stream.atEnd()) {
+        stream >> singleChar;
+        if (singleChar == QChar('/')) {
+        }
```
Comment (karliss) [inline src/widgets/DecompilerWidget.cpp:588]: Debug code.

**Pull request merged**